File: godotandroidble/app/src/main/java/com/duffrecords/godotandroidble/CyclingCadenceMeasurement.java
package com.duffrecords.godotandroidble;

import androidx.annotation.NonNull;

import com.welie.blessed.BluetoothBytesParser;

import java.io.Serializable;
import java.util.Locale;

import static com.welie.blessed.BluetoothBytesParser.FORMAT_UINT32;
import static com.welie.blessed.BluetoothBytesParser.FORMAT_UINT16;
import static com.welie.blessed.BluetoothBytesParser.FORMAT_UINT8;

public class CyclingCadenceMeasurement implements Serializable {

    public final Integer revs;
    public final Integer last;

    public CyclingCadenceMeasurement(byte[] value) {
        BluetoothBytesParser parser = new BluetoothBytesParser(value);

        // Parse the flags
        int flags = parser.getIntValue(FORMAT_UINT8);
        final int unit = flags & 0x01;

        // Parse cumulative revolutions
        this.revs = (unit == 0) ? parser.getIntValue(FORMAT_UINT16) : parser.getIntValue(FORMAT_UINT32);
        this.last = parser.getIntValue(FORMAT_UINT16);
    }

    @NonNull
    @Override
    public String toString() {
        return String.format(Locale.ENGLISH, "revs: %d\tlast: %d", revs, last);
    }
}

File: godotandroidble/app/src/main/java/com/duffrecords/godotandroidble/CyclingCadenceMeasurement.kt
package com.duffrecords.godotandroidble

import com.welie.blessed.BluetoothBytesParser
import java.nio.ByteOrder

data class CyclingCadenceMeasurement(
    val cumulativeWheelRevs: UInt?,
    val lastWheelEventTime: UShort?,
    val cumulativeCrankRevs: UShort?,
    val lastCrankEventTime: UShort?
) {
    override fun toString(): String {
        return "wheel revs: $cumulativeWheelRevs last wheel event: $lastWheelEventTime crank revs: $cumulativeCrankRevs last crank time: $lastCrankEventTime"
    }

    companion object {
        fun fromBytes(value: ByteArray): CyclingCadenceMeasurement? {
            val parser = BluetoothBytesParser(value, 0, ByteOrder.LITTLE_ENDIAN)

            try {
                val flags = parser.getUInt8()
                val wheelRevDataPresent = flags and 0x01u > 0u
                val crankRevDataPresent = flags and 0x10u > 0u
                val cumulativeWheelRevs = if (wheelRevDataPresent) parser.getUInt32() else null
                val lastWheelEventTime = if (wheelRevDataPresent) parser.getUInt16() else null
                val cumulativeCrankRevs = if (crankRevDataPresent) parser.getUInt16() else null
                val lastCrankEventTime = if (crankRevDataPresent) parser.getUInt16() else null

                return CyclingCadenceMeasurement(
                    cumulativeWheelRevs = cumulativeWheelRevs,
                    lastWheelEventTime = lastWheelEventTime,
                    cumulativeCrankRevs = cumulativeCrankRevs,
                    lastCrankEventTime = lastCrankEventTime
                )
            } catch (_: Exception) {
                return null
            }
        }
    }
}